cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.runtime.spark.distributed;
+
+import co.cask.cdap.app.runtime.spark.SparkRunnerClassLoader;
+import co.cask.cdap.common.lang.ClassLoaders;
+import com.google.common.base.Throwables;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class launches Spark YARN containers with classes loaded through the {@link SparkRunnerClassLoader}.
+ */
+public final class SparkContainerLauncher {
+
+  /**
+   * Launches the given main class. The main class will be loaded through the {@link SparkRunnerClassLoader}.
+   *
+   * @param mainClassName the main class to launch
+   * @param args arguments for the main class
+   */
+  @SuppressWarnings("unused")
+  public static void launch(String mainClassName, String[] args) {
+    List<URL> urls = ClassLoaders.getClassLoaderURLs(ClassLoader.getSystemClassLoader(), new ArrayList<URL>());
+
+    // Remove the URL that contains the given main classname to avoid infinite recursion.
+    // This is needed because we generate a class with the same main classname in order to intercept the main()
+    // method call from the container launch script.
+    URL resource = ClassLoader.getSystemClassLoader().getResource(mainClassName.replace('.', '/') + ".class");
+    if (resource == null) {
+      throw new IllegalStateException("Failed to find resource for main class " + mainClassName);
+    }
+
+    if (!urls.remove(getClassPathURL(mainClassName, resource))) {
+      throw new IllegalStateException("Failed to remove main class resource " + resource);
+    }
+
+    ClassLoader classLoader = new SparkRunnerClassLoader(urls.toArray(new URL[urls.size()]),
+                                                         ClassLoader.getSystemClassLoader().getParent(),
+                                                         false);
+    Thread.currentThread().setContextClassLoader(classLoader);
```

Comment:
Probably because this will just launch the spark job and exit? Perhaps a comment here would help?